### PR TITLE
Understandable error message for incompatible AnyStr arguments

### LINKED ISSUE #963: confusing error message when using AnyStr
```python
from typing import AnyStr

def f(x: AnyStr, y: AnyStr) -> AnyStr:
    print(x + y)

f('one', b'two')
```

This correctly fails to typecheck because the second arg is bytes, but the error message is hard to follow:

    test.py:6: error: Type argument 1 of "f" has incompatible value "object"

### PR DESCRIPTION
**Pull request by quartox**
Fixes #963

If a function has two or more AnyStr arguments then it requires the passed arguments to be all "str" or all "bytes". This is now explained in the error message and a note.

### CODE REVIEW

**Comment by lincolnq** [comment]
(Note: I'm not a maintainer of this project and in fact am rather new to it, so this is just one person's random thoughts). 

I don't think this is the right approach: `AnyStr` is a typing module concern, not a mypy concern. Thus this change is brittle (because typing may change), dependent on having to peek into the name in order to succeed at the optimization, and not general enough -- other type variables would not get the same treatment, even identically-declared ones with different names.

Though it might be much harder to execute, I would much prefer to see a change where mypy was smart enough to generalize this logic to any type variable which is reused across function args.

File: mypy/applytype.py [hunk 1]
```diff
@@ -62,3 +66,16 @@ def apply_generic_arguments(callable: CallableType, types: List[Type],
...
+
+def has_anystr_incompatible_args(arg_strings: Sequence[str], type: Type) -> bool:
+    """Determines if function has a problem with AnyStr arguments.
+
+    If the function has more than one AnyStr argument and the solver returns the object type,
+    then the function was passed both an "str" and "bytes" argument type.
+    """
+    if isinstance(type, Instance) and type.type.name() == 'object':
+        for string in arg_strings:
+            if 'AnyStr' in string:
```
Comment (ilevkivskyi) [inline mypy/applytype.py]: I think checking for string representation is too fragile.

File: mypy/applytype.py [hunk 2]
```diff
@@ -38,8 +38,12 @@ def apply_generic_arguments(callable: CallableType, types: List[Type],
                     types[i] = value
                     break
             else:
-                msg.incompatible_typevar_value(callable, i + 1, type, context)
-
+                arg_strings = tuple(msg.format(arg_type).replace('"', '')
+                                    for arg_type in callable.arg_types)
```
Comment (ilevkivskyi) [inline mypy/applytype.py]: I think you should check ``arg_type`` for the presence of a constrained type variable here instead of string representation below.

File: test-data/unit/check-functions.test [hunk 3]
```diff
@@ -2068,6 +2068,40 @@ def fn(
...
+# N: "AnyStr" arguments must be all "str" or all "bytes"
+
+[case testStarAnyStrIncompatibleArguments]
+from typing import TypeVar, Union
+AnyStr = TypeVar('AnyStr', bytes, str)
+def f(*x: AnyStr) -> None: pass
+f('a')
+f('a', 'b')
+f('a', b'b') # E: Type arguments of "f('AnyStr')" have incompatible values \
+# N: "AnyStr" arguments must be all "str" or all "bytes"
```
Comment (ilevkivskyi) [inline test-data/unit/check-functions.test]: Some ideas for additional tests:
* Situations where there are more than one type variable
* Situations where ``AnyStr`` also appears in the return type.

File: mypy/applytype.py (hunk 1 above)
Comment (quartox) [inline mypy/applytype.py]: I am trying to figure out how to use a type visitor to check for the types here (passing `callable.arg_types` instead of strings). Can you help point me to an example using type visitor to check for types?

Comment (ilevkivskyi) [inline mypy/applytype.py]: There are many of them (use ``grep -rn mypy -e "TypeVisitor"`` and/or ``grep -rn mypy -e "TypeQuery"`` to see them). One of the simplest and closest to what you need is probably the one in ``typeanal.get_type_var_names``. Another simple example is ``expandtype.ExpandTypeVisitor``.

**Comment by quartox** [comment]
I know there are some failing checks but I wanted to see if this is a good direction.

I am now covering all constrained types, but need more testing of return values.

The biggest remaining problem is with `*` arguments because they break the indexing for the function arguments. I added a hack that works only if the final argument is a `*` argument, but it doesn't work with `*` in the first position of the function arguments. I couldn't figure out how to check for a function with a `*` argument, but decrementing the index appears to fix the problem.

The indexing in general is weird (see tests for examples). It may not be worth showing the user the index in these cases when it could cause more confusion than it is worth.

I spent a fair amount of time grepping and reading about visitors but I couldn't see a clear example of how to use one without having more variables. I think the new version is better than the string checking, but not sure if it is robust enough.

**Comment by ilevkivskyi** [comment]
This looks like moving in the right direction. I will make a more detailed review later(probably tomorrow). In the meantime please fix the problem with typeshed, it shouldn't be part of this PR.

**Comment by quartox** [comment]
I did a checkout of https://github.com/python/mypy.git typeshed/ and it is still showing file changes. The original changes were brought in on a merge and I have no idea how or why it happened. Let me know if I can do anything else to get rid of the typeshed changes.

**Comment by gvanrossum** [comment]
Do you have your local clone of the mypy repo configured with an "origin" and an "upstream" remotes? (`git remote -v` will show, and `git remote --help` will help you understand how to configure this.)

Typically origin is your own fork and the default for push/pull, while upstream is the python/mypy repo. With such a setup you should be able to do something like `git fetch upstream`; `git rebase upstream/master`; `git push origin` to clear things up. But without the ability to look into and experiment with your git config I can't really help you further -- though stackoverflow probably has the answer (!).

**Comment by quartox** [comment]
Oh, god. I did rebase with upstream (https://github.com/python/mypy.git) master but it has created a monster pull request. I may not have time to fix it for the next few days because I am travelling, but I will try to get back to it soon.

**Comment by ilevkivskyi** [comment]
I tried to review the relevant files, but it is really hard, sorry. I will wait until you fix this.

**Comment by ilevkivskyi** [comment]
OK, this looks much better now. Will review this PR soon.

**Comment by quartox** [comment]
@ilevkivskyi Yeah, I still have no idea how that happened (I haven't done any rebasing before), but I just did a manual `git checkout upstream/master` for all the files that I didn't want to be in this PR. I still expect some tests to fail but the code itself should actually be readable now.

File: mypy/messages.py [hunk 4]
```diff
@@ -819,6 +819,21 @@ def incompatible_typevar_value(self, callee: CallableType, index: int,
...
+                                           constraints: Mapping[str, Sequence[str]],
+                                           context: Context) -> None:
+        for key, values in constraints.items():
+            self.fail('Type argument {} of {} has incompatible value'.format(
+                index, callable_name(callee)), context)
+            if len(values) == 2:
+                constraint_str = '{} or {}'.format(values[0], values[1])
+            elif len(values) > 3:
+                constraint_str = ', '.join(values[:-1]) + ', or ' + values[-1]
+            self.note('"{}" must be all one type: {}'.format(
```
Comment (ilevkivskyi) [inline mypy/messages.py]: This error message is better than existing but still quite confusing. Ca you make it less terse?
Maybe something in the spirit of "Arguments 1, 2, and 4 to 'my_function' should be either all 'str' or all 'bytes'. Got (str, str, bytes)."

File: test-data/unit/check-functions.test [hunk 5]
```diff
@@ -2125,3 +2125,82 @@ def i() -> List[Union[str, int]]:
...
+# N: "S" must be all one type: int or str
+f('1', '2', b'3', '4') # E: Type argument 2 of "f" has incompatible value \
+# N: "AnyStr" must be all one type: str or bytes
+f('1', '2', '3', b'4') # E: Type argument 2 of "f" has incompatible value \
+# N: "AnyStr" must be all one type: str or bytes
+f('1', 2, b'3', '4') # E: Type argument 1 of "f" has incompatible value \
+# N: "S" must be all one type: int or str \
+# E: Type argument 2 of "f" has incompatible value \
+# N: "AnyStr" must be all one type: str or bytes
+f(1, '2', '3', b'4') # E: Type argument 1 of "f" has incompatible value \
```
Comment (ilevkivskyi) [inline test-data/unit/check-functions.test]: Could you please make tests _slightly_ more "sparse"?

File: mypy/applytype.py [hunk 6]
```diff
@@ -62,3 +68,32 @@ def apply_generic_arguments(callable: CallableType, types: List[Type],
...
     )
+
+
+def get_incompatible_arg_constraints(arg_types: Sequence[Type], type: Type,
+                                     index: int) -> Dict[str, Tuple[str]]:
+    """Gets incompatible function arguments with the constrained types.
+
+    An example of a constrained type is AnyStr which must be all str or all byte.
+    """
+    constraints = {}  # type: Dict[str, Tuple[str]]
```
Comment (ilevkivskyi) [inline mypy/applytype.py]: Again, maybe ``Tuple[str, ...]``?

File: mypy/applytype.py [hunk 7]
```diff
@@ -62,3 +68,32 @@ def apply_generic_arguments(callable: CallableType, types: List[Type],
...
+def add_arg_constraints(constraints: Dict[str, Tuple[str]],
+                        arg_type: Type) -> Dict[str, Tuple[str]]:
+    if (isinstance(arg_type, TypeVarType) and
+            arg_type.values and
+            len(arg_type.values) > 1 and
+            arg_type.name not in constraints.keys()):
+        constraints[arg_type.name] = tuple(vals.type.name() for vals in arg_type.values)
+    elif isinstance(arg_type, UnionType):
+        for item in arg_type.items:
+            constraints = add_arg_constraints(constraints, item)
```
Comment (ilevkivskyi) [inline mypy/applytype.py]: The logic seems to be over-complicated here. You are basically just collecting a set of type variables affected by the error encountered. This should be separated from the error formatting logic (the latter should be kept in ``messages.py``).

File: test-data/unit/check-functions.test [hunk 8]
```diff
@@ -2125,3 +2125,82 @@ def i() -> List[Union[str, int]]:
...
+f('a')
+f('a', 'b')
+f('a', b'b') # E: Type argument 1 of "f" has incompatible value \
+# N: "AnyStr" must be all one type: str or bytes
+g(1, 'a')
+g(1, 'a', b'b') # E: Type argument 1 of "g" has incompatible value \
+# N: "AnyStr" must be all one type: str or bytes
+h('a', y=1)
+h('a', 'b', y=1)
+h('a', b'b', y=1) # E: Type argument 1 of "h" has incompatible value "object"
```
Comment (quartox) [inline test-data/unit/check-functions.test]: Can you point me in the right direction on how to decode `Callable.arg_kinds`? For this case it is just `[2,3]` which doesn't mean much to me. Can I compare this to `Context.arg_kinds`? In this case it is `[0,0,3]`.

The only other time this error occurs is on line 961 of `check-overloading.test` which I don't think is in scope for this pull request and line 233 of `check-typevar-values.text` which is actually passing an argument with type `object`.

File: mypy/applytype.py [hunk 9]
```diff
@@ -62,3 +68,32 @@ def apply_generic_arguments(callable: CallableType, types: List[Type],
...
+
+
+def get_incompatible_arg_constraints(arg_types: Sequence[Type], type: Type,
+                                     index: int) -> Dict[str, Tuple[str]]:
+    """Gets incompatible function arguments with the constrained types.
+
+    An example of a constrained type is AnyStr which must be all str or all byte.
+    """
+    constraints = {}  # type: Dict[str, Tuple[str]]
+    if isinstance(type, Instance) and type.type.name() == 'object':
```
Comment (quartox) [inline mypy/applytype.py]: Is there another simple check to see if the constraints are being violated? I write this as a pretty restrictive check which means the rest of the code won't be executed very often (I was worried about adding too much overhead which might be pre-mature optimization). Maybe see if `type.type.name() != arg_types[index].values`? But it seems like there must be a better solution.

File: mypy/messages.py (hunk 4 above)
Comment (quartox) [inline mypy/messages.py]: I think I can probably find the argument indices easily enough, but the types of the other arguments isn't available when constructing the message because the message builder only gets called when the second argument is analyzed (from what I can tell, please let me know if there is something I have overlooked). I can get `mypy.nodes.StrExpr` and `mypy.nodes.BytesExpr` from `context`, but I do not know how to safely get these values as `MessageBuilder.format` for both of these returns `object`.

File: mypy/applytype.py (hunk 7 above)
Comment (quartox) [inline mypy/applytype.py]: Is this specific to the union type check or the function in general? Using this function recursively appeared to be the easiest way to get the set of all possible types for one argument (since, in principle, the union could contain multiple unions as well).

File: test-data/unit/check-functions.test (hunk 8 above)
Comment (ilevkivskyi) [inline test-data/unit/check-functions.test]: They are described in ``nodes.py`` above ``class CallExpr...``. I copy here for convenience:
```python
# Positional argument
ARG_POS = 0  # type: int
# Positional, optional argument (functions only, not calls)
ARG_OPT = 1  # type: int
# *arg argument
ARG_STAR = 2  # type: int
# Keyword argument x=y in call, or keyword-only function arg
ARG_NAMED = 3  # type: int
# **arg argument
ARG_STAR2 = 4  # type: int
# In an argument list, keyword-only and also optional
ARG_NAMED_OPT = 5
```
If this will take time, then maybe we need to postpone this for a separate PR. I don't want this one to take long.

File: mypy/messages.py (hunk 4 above)
Comment (ilevkivskyi) [inline mypy/messages.py]: OK, then again we could postpone this to the next PR (probably the idea is to use ``types`` argument for ``apply_generic_arguments`` and ``map_actuals_to_formals`` from ``checkexpr.py``).

File: mypy/applytype.py (hunk 7 above)
Comment (ilevkivskyi) [inline mypy/applytype.py]: > Is this specific to the union type check or the function in general?

In general. My main concern here is that error formatting is mixed with the collection of information needed for the error.

File: mypy/applytype.py [hunk 10]
```diff
@@ -62,3 +68,36 @@ def apply_generic_arguments(callable: CallableType, types: List[Type],
...
+def get_incompatible_arg_constraints(msg: MessageBuilder,
+                                     arg_types: Sequence[Type],
+                                     type: Type,
+                                     index: int) -> Dict[str, Tuple[str, ...]]:
+    """Gets incompatible function arguments with the constrained types.
+
+    An example of a constrained type is AnyStr which must be all str or all byte.
+    """
+    constraints = {}  # type: Dict[str, Tuple[str, ...]]
+    print(index)
```
Comment (ilevkivskyi) [inline mypy/applytype.py]: You forgot a debugging print here.

File: mypy/applytype.py (hunk 9 above)
Comment (ilevkivskyi) [inline mypy/applytype.py]: > Is there another simple check to see if the constraints are being violated?

Yes, probably it would be quite complex. But then could you please make it more clear that we are special-casing ``AnyStr`` and similar? (This includes function names and more detailed docstrings.)

File: mypy/messages.py (hunk 4 above)
Comment (ilevkivskyi) [inline mypy/messages.py]: Still I think that error message like "Arguments 1, 2, and 4 to 'my_function' should be either all 'str' or all 'bytes'" will look better than the current one, so it is better to change it in this PR.

File: mypy/applytype.py [hunk 11]
```diff
@@ -38,8 +41,11 @@ def apply_generic_arguments(callable: CallableType, types: List[Type],
                     types[i] = value
                     break
             else:
-                msg.incompatible_typevar_value(callable, i + 1, type, context)
-
+                constraints = get_incompatible_arg_constraints(msg, callable.arg_types, type, i + 1)
```
Comment (ilevkivskyi) [inline mypy/applytype.py]: Could you please also fix the lint issue -- this line is too long (more than 99 allowed chars).

**Comment by ilevkivskyi** [comment]
@quartox Thanks for keeping forking on this! Just few more comments. Could you please also fix the tests (you forgot quotes "..." around the types in some tests)?

**Comment by quartox** [comment]
@ilevkivskyi I am running the tests, but I wanted to give you the chance to look at updates in case I don't fix tests before tomorrow.

**Comment by quartox** [comment]
I realized that my string checking for indeces is brittle. Need to rethink it.

**Comment by ilevkivskyi** [comment]
@quartox 
> I realized that my string checking for indeces is brittle. Need to rethink it. Any suggestions welcome.

TBH, I am a bit lost here. What is the current status? Have you addressed the comments from the previous review?

Also I am not sure what do you mean by "string checking for indeces". This PR takes too long, let us just see if we can get something mergeable from it without making an overly general solution.

**Comment by ilevkivskyi** [comment]
@quartox Could you please ping me when this will be ready for a next round of review?
(Note that the tests are still failing.)

**Comment by quartox** [comment]
I am fixing the tests this morning, it should be ready when the tests pass.

**Comment by ilevkivskyi** [comment]
After looking at this again, I am quite pessimistic, maybe we should just take a different approach. I have unassigned myself, in case if someone else wants to try pushing this forward. Sorry!

**Comment by quartox** [comment]
@ilevkivskyi thanks for the help. I think I agree, this might be better with a fresh approach, but if another reviewer has other suggestions I am happy to help improve.

**Comment by gvanrossum** [comment]
Sadly I think I have to agree with Ivan. Thanks for all your efforts nevertheless! Maybe you are still interested in contributing to other parts of mypy? (In general I would recommend starting on smaller issues -- maybe you have a better appreciation for what that means now that you've worked on this for a while.)
